### ISSUE TITLE
Could not deduce ‘MonadFix m’

### ISSUE THREAD

**Issue by guibou** [DanBurton/tardis#15]
When compiling with GHC 9.6, I'm having the following compilation error:

```
src/Control/Monad/Trans/Tardis.hs:136:10: error: [GHC-39999]
    • Could not deduce ‘MonadFix m’
        arising from the head of a quantified constraint
        arising from the superclasses of an instance declaration
      from the context: Monad m
        bound by a quantified context
        at src/Control/Monad/Trans/Tardis.hs:136:10-35
      Possible fix:
        add (MonadFix m) to the context of a quantified context
    • In the instance declaration for ‘MonadTrans (TardisT bw fw)’
    |
136 | instance MonadTrans (TardisT bw fw) where
```

From my understanding, the problem comes from the fact tha GHC 9.6 comes with `transformers-0.6.1.0` which had this new requiredement for [MonadTrans`](https://hackage.haskell.org/package/transformers-0.6.1.0/docs/Control-Monad-Trans-Class.html#t:MonadTrans)

```
class (forall m. Monad m => Monad (t m)) => MonadTrans t where
```

> Since 0.6.0.0 and for GHC 8.6 and later, the requirement that t m be a [Monad](https://hackage.haskell.org/package/base-4.16.3.0/docs/Control-Monad.html#t:Monad) is enforced by the implication constraint forall m. [Monad](https://hackage.haskell.org/package/base-4.16.3.0/docs/Control-Monad.html#t:Monad) m => [Monad](https://hackage.haskell.org/package/base-4.16.3.0/docs/Control-Monad.html#t:Monad) (t m) enabled by the QuantifiedConstraints extension.

It means that `instance MonadTrans (TardisT bw fw)` must satisfy `Monad (TardisT bw fw m)` (because `lift :: Monad m => m a -> TardisT bw fw m a`. (That's my understanding on how this `=>` works, I may be wrong).

However, it implies `MonadFix m`, because of the instance `instance MonadFix m => Monad (TardisT bw fw m)`.

So we need to provide a proof in the instance declaration that `MonadFix m` is true for all `m`, which is as far as I know, impossible. It may be related to #8.

So I'm wondering what can be the solution to compile the tardis package with ghc 9.6? The obvious solution would be to remove the `TardisT`, or actually, remove the `MonadTrans` instance and let user lift using:

```
liftTardis :: Monad m => m a -> TardisT bw fw m a
liftTardis m = TardisT $ \s -> do
  x <- m
  return (x, s)
```

**Comment by guibou** [comment]
@Lysxia 

```
instance MonadFix m => Monad (TardisT bw fw m) where
  return x = tardis $ \s -> (x, s)
  m >>= f  = TardisT $ \ ~(bw, fw) -> do
    rec (x,  ~(bw'', fw' )) <- runTardisT m (bw', fw)
        (x', ~(bw' , fw'')) <- runTardisT (f x) (bw, fw')
    return (x', (bw'', fw''))
```

I don't think I can remove the `MonadFix m` here. Because if I passe `mfix` in parameter somewhere here, I will need to either have `MonadFix m` as constraint, OR I need a way to pass `mfix` as parameter to `return` and `>>=`, which will change the `Monad` api. Unless I missed something.

**Comment by Lysxia** [comment]
I was proposing one of the following implementations:

```haskell
{-# LANGUAGE RecursiveDo, RankNTypes #-}
module Tardis where

import Control.Monad
import Control.Monad.Fix
import Data.Bifunctor

-- Version 1

newtype TardisT bw fw m a = TardisT { runTardisT :: MonadFix m => (bw, fw) -> m (a, (bw, fw)) }

instance Monad (TardisT bw fw m) where
  return x = TardisT $ \s -> pure (x, s)
  m >>= f  = TardisT $ \ ~(bw, fw) -> do
    rec (x,  ~(bw'', fw' )) <- runTardisT m (bw', fw)
        (x', ~(bw' , fw'')) <- runTardisT (f x) (bw, fw')
    return (x', (bw'', fw''))

instance Functor (TardisT bw fw m) where fmap = liftM
instance Applicative (TardisT bw fw m) where pure = return ; (<*>) = ap


-- Version 2

newtype TardisT' bw fw m a = TardisT' { unTardisT' :: (forall a. (a -> m a) -> m a) -> (bw, fw) -> m (a, (bw, fw)) }

runTardisT' :: MonadFix m => TardisT' bw fw m a -> (bw, fw) -> m (a, (bw, fw))
runTardisT' (TardisT' f) = f mfix

instance Monad m => Monad (TardisT' bw fw m) where
  return x = TardisT' $ \_ s -> pure (x, s)
  m >>= f  = TardisT' $ \mfix ~(bw, fw) ->
    fst <$> mfix (\ ~(_, bw'_) -> do
      (x, ~(bw'', fw')) <- unTardisT' m mfix (bw'_, fw)
      (y, ~(bw' , fw'')) <- unTardisT' (f x) mfix (bw, fw')
      pure ((y, (bw'', fw'')), bw'))

instance Functor m => Functor (TardisT' bw fw m) where
  fmap f (TardisT' u) = TardisT' (\mfix -> (fmap . fmap . first) f (u mfix))

instance Monad m => Applicative (TardisT' bw fw m) where
  pure = return ; (<*>) = ap
```

**Comment by guibou** [comment]
That's great, I was initially unable to get this to work because I implemented `return` using `tardis` which needs a `MonadFix` constraint.

This builds, you are right.

I'm again lost with the way I understand constraints. `TardisT { runTardisT :: MonadFix m => (bw, fw) -> m (a, (bw, fw)) }` does mean that *when* `runTardisT` is called, it must be called in a context where `MonadFix m` holds, but it says nothing about what should be `m` when `TardisT` is called. This puzzles me, always.

Now the problem is with `mapTardisT`:

```

src/Control/Monad/Trans/Tardis.hs:110:32: error: [GHC-39999]
    • Could not deduce ‘MonadFix m’ arising from a use of ‘runTardisT’
      from the context: MonadFix n
        bound by a type expected by the context:
                   MonadFix n => (bw, fw) -> n (b, (bw, fw))
        at src/Control/Monad/Trans/Tardis.hs:110:28-43
      Possible fix:
        add (MonadFix m) to the context of
          the type signature for:
            mapTardisT :: forall (m :: * -> *) a bw fw (n :: * -> *) b.
                          (m (a, (bw, fw)) -> n (b, (bw, fw)))
                          -> TardisT bw fw m a -> TardisT bw fw n b
    • In the second argument of ‘(.)’, namely ‘runTardisT m’
      In the second argument of ‘($)’, namely ‘f . runTardisT m’
      In the expression: TardisT $ f . runTardisT m
    |
110 | mapTardisT f m = TardisT $ f . runTardisT m
```

And if I add `MonadFix m` to the context of `mapTardisT`, I do now have a problem with `MFunctor`:

```
src/Control/Monad/Trans/Tardis.hs:149:13: error: [GHC-39999]
    • Could not deduce ‘MonadFix m’ arising from a use of ‘mapTardisT’
      from the context: Monad m
        bound by the type signature for:
                   hoist :: forall (m :: * -> *) (n :: * -> *) b.
                            Monad m =>
                            (forall a. m a -> n a) -> TardisT bw fw m b -> TardisT bw fw n b
        at src/Control/Monad/Trans/Tardis.hs:149:3-7
      Possible fix:
        add (MonadFix m) to the context of
          the type signature for:
            hoist :: forall (m :: * -> *) (n :: * -> *) b.
                     Monad m =>
                     (forall a. m a -> n a) -> TardisT bw fw m b -> TardisT bw fw n b
    • In the expression: mapTardisT f
      In an equation for ‘hoist’: hoist f = mapTardisT f
      In the instance declaration for ‘MFunctor (TardisT bw fw)’
    |
149 |   hoist f = mapTardisT f
```

We are making progress. Looks like it is possible to keep the instance for `MonadTrans` and remove the one from `MFunctor`.

**Comment by alaendle** [comment]
(just for reference - also found on stackage - https://github.com/commercialhaskell/stackage/issues/7017#issuecomment-1601960987)

**Comment by guibou** [comment]
This was fixed in `tardis` `0.5.0` by removing the `MonadTrans` instance. Thank you @DanBurton.

**Comment by DanBurton** [comment]
> The obvious solution would be... remove the `MonadTrans` instance and let user `lift` using `liftTardis`

I did indeed decide to go this route. I called it `liftTardisT`. Thank you for the suggestion.
